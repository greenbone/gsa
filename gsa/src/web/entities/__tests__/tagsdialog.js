/* Copyright (C) 2019-2020 Greenbone Networks GmbH
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react';

import {render, fireEvent} from 'web/utils/testing';

import Dialog from '../tagsdialog';

describe('TagsDialog dialog component tests', () => {
  test('should render dialog', () => {
    const handleClose = jest.fn();
    const handleNewTagClick = jest.fn();
    const handleSave = jest.fn();
    const handleTagChange = jest.fn();

    const {baseElement} = render(
      <Dialog
        entitiesCount={3}
        tags={[{name: 'foo', id: '123'}]}
        onClose={handleClose}
        onNewTagClick={handleNewTagClick}
        onSave={handleSave}
        onTagChanged={handleTagChange}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('should disable tag selection when no options available', () => {
    const handleClose = jest.fn();
    const handleNewTagClick = jest.fn();
    const handleSave = jest.fn();
    const handleTagChange = jest.fn();

    const {baseElement} = render(
      <Dialog
        entitiesCount={3}
        onClose={handleClose}
        onNewTagClick={handleNewTagClick}
        onSave={handleSave}
        onTagChanged={handleTagChange}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('should save data', () => {
    const handleClose = jest.fn();
    const handleNewTagClick = jest.fn();
    const handleSave = jest.fn();
    const handleTagChange = jest.fn();

    const {getByTestId} = render(
      <Dialog
        comment="foo"
        entitiesCount={3}
        name="bar"
        tagId="123"
        tags={[{name: 'foo', id: '123'}]}
        value="lorem"
        onClose={handleClose}
        onNewTagClick={handleNewTagClick}
        onSave={handleSave}
        onTagChanged={handleTagChange}
      />,
    );

    const saveButton = getByTestId('dialog-save-button');
    fireEvent.click(saveButton);
    expect(handleSave).toHaveBeenCalledWith({
      comment: 'foo',
      id: '123',
      name: 'bar',
      value: 'lorem',
    });
  });

  test('should allow to close the dialog', () => {
    const handleClose = jest.fn();
    const handleNewTagClick = jest.fn();
    const handleSave = jest.fn();
    const handleTagChange = jest.fn();

    const {getByTestId} = render(
      <Dialog
        entitiesCount={3}
        tags={[{name: 'foo', id: '123'}]}
        onClose={handleClose}
        onNewTagClick={handleNewTagClick}
        onSave={handleSave}
        onTagChanged={handleTagChange}
      />,
    );

    const closeButton = getByTestId('dialog-close-button');

    fireEvent.click(closeButton);

    expect(handleClose).toHaveBeenCalled();
  });
});
