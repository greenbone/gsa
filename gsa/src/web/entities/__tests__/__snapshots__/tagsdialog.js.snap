// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsDialog dialog component tests should disable tag selection when no options available 1`] = `
<body>
  <div
    id="portals"
  >
    .c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -5px;
}

.c13 > * {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c13 > * {
  margin-left: 5px;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c20 {
  cursor: pointer;
}

.c21 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c21 * {
  height: inherit;
  width: inherit;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10% auto;
  border: 0;
  outline: 0;
  width: 650px;
  height: px;
}

.c0 {
  position: fixed;
  font-family: Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;
  font-size: 1.1em;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 0;
  background: rgba(102,102,102,0.5);
  z-index: 600;
  -webkit-transition: opacity 1s ease-in;
  transition: opacity 1s ease-in;
  width: 100%;
  height: 100%;
}

.c27 {
  width: 0;
  height: 0;
  cursor: nwse-resize;
  border-right: 20px solid transparent;
  border-bottom: 20px solid transparent;
  border-top: 20px solid #fff;
}

.c26 {
  position: absolute;
  bottom: 3px;
  right: 3px;
  width: 20px;
  height: 20px;
  background: repeating-linear-gradient( -45deg, transparent, transparent 2px, #000 2px, #000 3px );
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: inherit;
  padding: 0;
  background: #fff;
  box-shadow: 5px 5px 10px #787878;
  border-radius: 3px;
  border: 1px solid #787878;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 1px solid #519032;
  font-weight: bold;
  font-size: 12px;
  font-family: Verdana,sans-serif;
  color: #519032;
  cursor: pointer;
  border-radius: 2px;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4:hover {
  color: #fff;
  background: #519032;
}

.c5 {
  height: 24px;
  width: 24px;
  line-height: 24px;
}

.c5 * {
  height: inherit;
  width: inherit;
}

.c23 {
  display: inline-block;
  padding: 0 15px;
  color: #393637;
  text-align: center;
  vertical-align: middle;
  font-size: 11px;
  font-weight: bold;
  line-height: 30px;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background-color: #fff;
  border-radius: 2px;
  border: 1px solid #aaaaaa;
  cursor: pointer;
  overflow: visible;
  z-index: 1;
}

.c23:focus,
.c23:hover {
  border: 1px solid #393637;
}

.c23:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #66c430;
  font-weight: bold;
  color: #fff;
}

.c23[disabled] {
  cursor: not-allowed;
  opacity: 0.65;
  box-shadow: none;
}

.c23 img {
  height: 32px;
  width: 32px;
  margin-top: 5px 10px 5px -10px;
  vertical-align: middle;
}

.c23:link {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #393637;
}

.c24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c25 {
  border: 1px solid #787878;
  color: #519032;
  background: #87d050;
}

.c25:hover {
  color: #fff;
  background: #519032;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-width: 1px 0 0 0;
  border-style: solid;
  border-color: #c8d3d9;
  margin-top: 15px;
  padding: 10px 20px 10px 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 5px 5px 10px;
  margin-bottom: 15px;
  border-radius: 2px 2px 0 0;
  border-bottom: 1px solid #787878;
  color: #fff;
  font-weight: bold;
  background: #66c430;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
}

.c7 {
  overflow: auto;
  padding: 0 15px;
  width: 100%;
  height: 100%;
  max-height: 400px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: hidden;
  height: 100%;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  padding-bottom: 10px;
}

.c10 {
  display: inline-block;
  max-width: 100%;
  font-weight: bold;
  text-align: right;
  padding-left: 10px;
  padding-right: 10px;
  width: 33.33333333%;
  margin-left: 0;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 66.66666667%;
  padding-left: 10px;
  padding-right: 10px;
}

.c18 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  margin: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c19 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c19 * {
  height: inherit;
  width: inherit;
}

.c15 {
  border: 1px solid #aaaaaa;
  border-radius: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 1px 5px;
  background-color: #fff;
  color: #000;
  font-weight: normal;
  background-color: #eeeeee;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 230px;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  word-break: keep-all;
  white-space: nowrap;
  overflow: hidden;
  cursor: default;
  cursor: default;
}

@media print {
  .c20 {
    display: none;
  }
}

<div>
      <div
        class="c0"
      >
        <div
          class="c1"
          height="px"
          role="dialog"
          tabindex="1"
          width="650px"
        >
          <div
            class="c2"
          >
            <div
              class="c3"
              data-testid="dialog-title-bar"
            >
              <span>
                Add Tag
              </span>
              <div
                class="c4 c5"
                data-testid="dialog-title-close-button"
                title="Close"
              >
                ×
              </div>
            </div>
            <div
              class="c6"
            >
              <div
                class="c7"
                data-testid="save-dialog-content"
              >
                <div
                  class="c8"
                >
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Choose Tag
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    >
                      <div
                        class="c12"
                      >
                        <div
                          class="c13"
                          margin="5px"
                        >
                          <div
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-labelledby="downshift-1-label"
                            class="c14"
                            role="combobox"
                            width="230px"
                          >
                            <div
                              aria-haspopup="true"
                              aria-label="open menu"
                              class="c15"
                              data-toggle="true"
                              disabled=""
                              role="button"
                              type="button"
                            >
                              <div
                                class="c16"
                                data-testid="select-selected-value"
                                disabled=""
                              />
                              <div
                                class="c17"
                              >
                                <span
                                  class="c18 c19"
                                  data-testid="select-open-button"
                                >
                                  ▼
                                </span>
                              </div>
                            </div>
                          </div>
                          <span
                            class="c20 c21"
                            data-testid="svg-icon"
                            title="Create a new Tag"
                          >
                            <svg
                              title="Create a new Tag"
                            >
                              new.svg
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Value
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    />
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Comment
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="c22"
            >
              <button
                class="c23 c24 c25"
                data-testid="dialog-close-button"
                title="Cancel"
              >
                Cancel
              </button>
              <button
                class="c23 c24 c25"
                data-testid="dialog-save-button"
                title="Add Tag"
              >
                Add Tag
              </button>
            </div>
          </div>
          <div
            class="c26"
          >
            <div
              class="c27"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`TagsDialog dialog component tests should render dialog 1`] = `
.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -5px;
}

.c13 > * {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c13 > * {
  margin-left: 5px;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c20 {
  cursor: pointer;
}

.c21 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c21 * {
  height: inherit;
  width: inherit;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10% auto;
  border: 0;
  outline: 0;
  width: 650px;
  height: px;
}

.c0 {
  position: fixed;
  font-family: Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;
  font-size: 1.1em;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 0;
  background: rgba(102,102,102,0.5);
  z-index: 600;
  -webkit-transition: opacity 1s ease-in;
  transition: opacity 1s ease-in;
  width: 100%;
  height: 100%;
}

.c27 {
  width: 0;
  height: 0;
  cursor: nwse-resize;
  border-right: 20px solid transparent;
  border-bottom: 20px solid transparent;
  border-top: 20px solid #fff;
}

.c26 {
  position: absolute;
  bottom: 3px;
  right: 3px;
  width: 20px;
  height: 20px;
  background: repeating-linear-gradient( -45deg, transparent, transparent 2px, #000 2px, #000 3px );
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: inherit;
  padding: 0;
  background: #fff;
  box-shadow: 5px 5px 10px #787878;
  border-radius: 3px;
  border: 1px solid #787878;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 1px solid #519032;
  font-weight: bold;
  font-size: 12px;
  font-family: Verdana,sans-serif;
  color: #519032;
  cursor: pointer;
  border-radius: 2px;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4:hover {
  color: #fff;
  background: #519032;
}

.c5 {
  height: 24px;
  width: 24px;
  line-height: 24px;
}

.c5 * {
  height: inherit;
  width: inherit;
}

.c23 {
  display: inline-block;
  padding: 0 15px;
  color: #393637;
  text-align: center;
  vertical-align: middle;
  font-size: 11px;
  font-weight: bold;
  line-height: 30px;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background-color: #fff;
  border-radius: 2px;
  border: 1px solid #aaaaaa;
  cursor: pointer;
  overflow: visible;
  z-index: 1;
}

.c23:focus,
.c23:hover {
  border: 1px solid #393637;
}

.c23:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #66c430;
  font-weight: bold;
  color: #fff;
}

.c23[disabled] {
  cursor: not-allowed;
  opacity: 0.65;
  box-shadow: none;
}

.c23 img {
  height: 32px;
  width: 32px;
  margin-top: 5px 10px 5px -10px;
  vertical-align: middle;
}

.c23:link {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #393637;
}

.c24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c25 {
  border: 1px solid #787878;
  color: #519032;
  background: #87d050;
}

.c25:hover {
  color: #fff;
  background: #519032;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-width: 1px 0 0 0;
  border-style: solid;
  border-color: #c8d3d9;
  margin-top: 15px;
  padding: 10px 20px 10px 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 5px 5px 10px;
  margin-bottom: 15px;
  border-radius: 2px 2px 0 0;
  border-bottom: 1px solid #787878;
  color: #fff;
  font-weight: bold;
  background: #66c430;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: -webkit-grab;
  cursor: -moz-grab;
  cursor: grab;
}

.c7 {
  overflow: auto;
  padding: 0 15px;
  width: 100%;
  height: 100%;
  max-height: 400px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: hidden;
  height: 100%;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  padding-bottom: 10px;
}

.c10 {
  display: inline-block;
  max-width: 100%;
  font-weight: bold;
  text-align: right;
  padding-left: 10px;
  padding-right: 10px;
  width: 33.33333333%;
  margin-left: 0;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 66.66666667%;
  padding-left: 10px;
  padding-right: 10px;
}

.c18 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  margin: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c19 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c19 * {
  height: inherit;
  width: inherit;
}

.c15 {
  border: 1px solid #aaaaaa;
  border-radius: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 1px 5px;
  background-color: #fff;
  color: #000;
  font-weight: normal;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 230px;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  word-break: keep-all;
  white-space: nowrap;
  overflow: hidden;
  cursor: pointer;
  cursor: default;
}

@media print {
  .c20 {
    display: none;
  }
}

<body>
  <div
    id="portals"
  >
    <div>
      <div
        class="c0"
      >
        <div
          class="c1"
          height="px"
          role="dialog"
          tabindex="1"
          width="650px"
        >
          <div
            class="c2"
          >
            <div
              class="c3"
              data-testid="dialog-title-bar"
            >
              <span>
                Add Tag
              </span>
              <div
                class="c4 c5"
                data-testid="dialog-title-close-button"
                title="Close"
              >
                ×
              </div>
            </div>
            <div
              class="c6"
            >
              <div
                class="c7"
                data-testid="save-dialog-content"
              >
                <div
                  class="c8"
                >
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Choose Tag
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    >
                      <div
                        class="c12"
                      >
                        <div
                          class="c13"
                          margin="5px"
                        >
                          <div
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-labelledby="downshift-0-label"
                            class="c14"
                            role="combobox"
                            width="230px"
                          >
                            <div
                              aria-haspopup="true"
                              aria-label="open menu"
                              class="c15"
                              data-toggle="true"
                              role="button"
                              type="button"
                            >
                              <div
                                class="c16"
                                data-testid="select-selected-value"
                              />
                              <div
                                class="c17"
                              >
                                <span
                                  class="c18 c19"
                                  data-testid="select-open-button"
                                >
                                  ▼
                                </span>
                              </div>
                            </div>
                          </div>
                          <span
                            class="c20 c21"
                            data-testid="svg-icon"
                            title="Create a new Tag"
                          >
                            <svg
                              title="Create a new Tag"
                            >
                              new.svg
                            </svg>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Value
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    />
                  </div>
                  <div
                    class="c9"
                  >
                    <label
                      class="c10"
                      data-testid="formgroup-title"
                    >
                      Comment
                    </label>
                    <div
                      class="c11"
                      data-testid="formgroup-content"
                      size="8"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="c22"
            >
              <button
                class="c23 c24 c25"
                data-testid="dialog-close-button"
                title="Cancel"
              >
                Cancel
              </button>
              <button
                class="c23 c24 c25"
                data-testid="dialog-save-button"
                title="Add Tag"
              >
                Add Tag
              </button>
            </div>
          </div>
          <div
            class="c26"
          >
            <div
              class="c27"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
